restock-shop.lic

```diff
@@ -12,98 +12,95 @@ class ShopRestock
-      ['gloves', 60_000, 4, 'in cabinet', 4, 'quilted cloth gloves', 1, 1, 'heavy burlap'],
-      ['shirt', 220_000, 34, 'in cabinet', 4, 'quilted cloth shirt', 0, 2, 'heavy burlap'],
-      ['pants', 120_000, 10, 'in cabinet', 4, 'quilted cloth pants', 1, 1, 'heavy burlap'],
-      ['backpack', 100_000, 5, 'on table', 3, 'backpack', 'heavy linen'],
-      ['bag', 100_000, 6, 'on table', 3, 'duffel bag', 'heavy linen'],
-      ['sealed', 220_000, 8, 'in cabinet', 'gryphon'],
-      ['sword', 140_000, 9, 'on rack', 'bastard sword', 'steel'],
-      ['spear', 120_000, 7, 'on rack', 'light metal spear', 'steel'],
-      ['mace', 140_000, 12, 'on rack', 'bar mace', 'steel'],
-      ['cutlass', 120_000, 7, 'on rack', 'cutlass', 'steel'],
-      ['bola', 120_000, 5, 'on rack', 'bola', 'steel'],
-      ['hammer', 120_000, 10, 'on rack', 'throwing hammer', 'steel'],
-      ['akabo', 120_000, 13, 'on rack', 'akabo', 'steel'],
-      ['spike', 100_000, 4, 'on rack', 'throwing spike', 'steel'],
-      ['nightstick', 120_000, 7, 'on rack', 'nightstick', 'steel'],
-      ['greataxe', 120_000, 12, 'on rack', 'greataxe', 'steel'],
-      ['carryall', 200_000, 5, 'on table', 3, 'carryall', 'heavy silk'],
-      ['haversack', 200_000, 6, 'on table', 3, 'haversack', 'heavy silk']
-    ]
-
-    walk_to(@shop_id)
-    fput('open entrance')
-    pause 0.5
-    fput('go entrance')
-
+    @settings = get_settings
+
+    @shop_id = @settings.restock_shop_outside_room
+    @inner_id = @settings.restock_shop_inside_room
+    @entrance = @settings.restock_shop_entrance_noun
+    @item_list = @settings.restock_shop_items
+    @town = @settings.restock_shop_town
+    @part_data = get_data('crafting')[:recipe_parts]
+    @bag = @settings.crafting_container
+    # TODO: Validate users item list against recipe data. Ensure that there is only one match per user specified recipe.
+    # TODO: Validate users town. Allow only towns with plazas for simplicity
+
+    # @item_list = [
+    #   ['backpack', 100_000, 5, 'on table', 3, 'backpack', 'heavy linen'],
+    #   ['bag', 100_000, 6, 'on table', 3, 'duffel bag', 'heavy linen'],
+    #   ['sealed', 220_000, 8, 'in cabinet', 'gryphon'],
+    #   ['sword', 140_000, 9, 'on rack', 'bastard sword', 'steel'],
+    #   ['mace', 140_000, 12, 'on rack', 'bar mace', 'steel'],
+    #   ['hammer', 120_000, 10, 'on rack', 'throwing hammer', 'steel'],
+    #   ['greataxe', 120_000, 12, 'on rack', 'greataxe', 'steel'],
+    #   ['carryall', 200_000, 5, 'on table', 3, 'carryall', 'heavy silk'],
+    #   ['haversack', 200_000, 6, 'on table', 3, 'haversack', 'heavy silk']
+    # ]
+    go_to_shop
     check_surfaces
 
-    return if Room.current.id == @inner_id
-    walk_to(@shop_id)
-    fput('open entrance')
-    pause 0.5
-    fput('go entrance')
+    go_to_shop
   end
 
   def check_surfaces
     missing_items = []
-    @item_list.map { |x| x[3] }.uniq.each do |surface|
+    @item_list.map { |data| data['surface'] }.uniq.each do |surface|
```

Comment:
💡 I wonder if this could be `missing_items = @item_list...uniq.map do |surface|` and then remove `missing_items +=`  from line 53. You may need a `flatten` at the end.

Reply:
💡 Or, maybe this is better...
```ruby
missing_items = @item_list
.group_by { |i| i['surface'] }
.map do |surface, items|
    raw_list =...
    items.reject { |item| raw_list.find { |raw_item| raw_item.match(item['full_name']) } }
end
```